Document Calico CNI tweaks (#258)
submariner-io/submariner-website: src/content/deployment/calico/_index.en.md

+++
title =  "Calico CNI"
weight = 30
+++

> Normally, the local CNI in your K8s cluster (based on Kube-proxy with iptables mode) programs iptable
> rules for networking of the Pods. When a Pod in your cluster tries to talk to an external-IP, the CNI
> generally performs MASQ/SNAT of the traffic as it does not belong to the local cluster.
> Similar to CNI, Submariner also programs certain IPTable rules and it requires these rules to be applied
> prior to the CNI rules in order to avoid the SNAT for inter-cluster traffic.
> Submariner tries to preserve the source-IP of the Pods for cross-cluster communication for visibility,
> ease of debugging, and for network policy support.

Reviewer: ```suggestion
Typically, the Kubernetes network plugin (based on kube-proxy) programs iptables
rules for Pods networking within the cluster. When a Pod in the cluster tries to access an external IP, the plugin
performs different Network Address Translation (NAT) manipulation on the traffic as it does not belong to the local cluster. Similarly, Submariner also programs certain iptables rules and it requires these rules to be applied
prior to the ones programmed by the network plugin.
Submariner tries to preserve the source IP of the Pods for cross-cluster communication for visibility,
ease of debugging, and security purposes.
```

Author: Done

> So, Submariner requires its iptable rules have higher precedence over the iptable rules that are programmed
> by the local Cluster CNI. With Calico, it was seen that Calico is pretty aggressive in inserting its rules
> at the top and this was causing issues with cross-cluster communication.
> However, we have a solution. We can instruct Calico, via IPPools, not to perform NAT'ing on certain subnets.
> In our case, its the Pod and Service CIDRs of the remote Clusters. Once the IPPools are configured in the
> clusters, Calico will not perform SNAT for the configured CIDRs and allows Submariner to support cross-cluster
> connectivity.

Reviewer: ```suggestion
So, Submariner requires its iptables rules to take precedence over the rules that are programmed
by the local cluster network plugin. On clusters deployed with Calico as the network plugin, it was seen that rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster communication.
To make Calico compatible with Submariner, it needs to be configured, via IPPools, not to perform NAT on the subnets associated with the Pod and Service CIDRs of the remote clusters. Once the IPPools are configured in the
clusters, Calico will not perform NAT for the configured CIDRs and allows Submariner to support cross-cluster
connectivity.
```

Author: Done

{{% notice note %}}
When using [Submariner Globalnet](../../architecture/globalnet) with Calico, please avoid the default
globalnet-cidr (i.e., 169.254.0.0/16) as its used internally within Calico. You can explicitly specify
a non-overlapping globalnet-cidr while deploying Submariner.
{{% /notice %}}

**Let's take an example**:

> If you have two clusters, East and West, with Calico CNI connected via Submariner, you can create the following IPPools.

Reviewer: ```suggestion
If you have two clusters, East and West, deployed with the Calico network plugin and connected via Submariner, you should create the following IPPools:
```

+++
title =  "Calico CNI"
weight = 30
+++

> Typically, the Kubernetes network plugin (based on kube-proxy) programs iptables rules for Pods

Author: Done

> networking within the cluster. When a Pod in the cluster tries to access an external IP, the plugin

Author: Done

performs different Network Address Translation (NAT) manipulation on the traffic as it does not
belong to the local cluster. Similarly, Submariner also programs certain iptables rules and it
requires these rules to be applied prior to the ones programmed by the network plugin.
Submariner tries to preserve the source IP of the Pods for cross-cluster communication for visibility,
ease of debugging, and security purposes.

> So, Submariner requires its iptables rules to take precedence over the rules that are programmed
> by the local cluster network plugin. On clusters deployed with Calico as the network plugin, it was
> seen that rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster

Reviewer 2: ```suggestion
On clusters deployed with Calico as the network plugin, the rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster
```

communication. To make Calico compatible with Submariner, it needs to be configured, via IPPools,
not to perform NAT on the subnets associated with the Pod and Service CIDRs of the remote clusters.
Once the IPPools are configured in the clusters, Calico will not perform NAT for the configured CIDRs
and allows Submariner to support cross-cluster connectivity.

{{% notice note %}}
When using [Submariner Globalnet](../../architecture/globalnet) with Calico, please avoid the default
globalnet-cidr (i.e., 169.254.0.0/16) as its used internally within Calico. You can explicitly specify
a non-overlapping globalnet-cidr while deploying Submariner.
{{% /notice %}}

> **Let's take an example**:
> If you have two clusters, East and West, deployed with the Calico network plugin and connected
> via Submariner, you should create the following IPPools:

Reviewer 2: ```suggestion
As an example, consider two clusters, East and West, deployed with the Calico network plugin and connected via Submariner. For cluster East, the Service CIDR is 100.93.0.0/16 and the Pod CIDR is 
10.243.0.0/16. For cluster West, they are 100.92.0.0/16 and 10.242.0.0/16. The following IPPools
should be created:
```
